broker-daemon/interchain/index.js

```diff
@@ -0,0 +1,275 @@
+ *                                            payment to to retrieve the
+ *                                            preimage.
+ * @returns {PaymentResult}                   Result of retrieving the preimage.
+ */
+async function getPreimage (
+  hash,
+  { engine: inboundEngine },
+  { engine: outboundEngine, amount: outboundAmount, address: outboundAddress }
+) {
+  // We don't want to reject an incoming HTLC if we know that there is an
+  // active outgoing one. If the outgoing HTLC is in flight or completed, we
+  // should attempt to retrieve the associated preimage.
+  logger.debug(`Checking outbound HTLC status for swap ${hash}`)
+  if (await outboundEngine.isPaymentPendingOrComplete(hash)) {
+    // TODO: return permanent errors when encountered
+    return {
+      paymentPreimage: await outboundEngine.getPaymentPreimage(hash)
+    }
+  }
+
+  // Subscribe to incoming payments that are accepted, but not yet settled.
+  try {
+    // TODO: implement this method
+    await inboundEngine.waitForSwapCommitment(hash)
+  } catch (e) {
+    // TODO: fix this error message
+    if (e.isSettled) {
+      logger.debug(`Swap for ${hash} has already been settled`)
+
+      return {
+        // TODO: update this method to be a simple request/response
+        paymentPreimage: await inboundEngine.getSettledSwapPreimage(hash)
+      }
+    }
+
+    throw e
+  }
+
+  logger.debug(`Sending payment to ${outboundAddress} to translate ${hash}`, {
+    outboundTimeLock: OUTBOUND_TIME_LOCK,
+    outboundAmount
+  })
+
+  // TODO: update this method to take a max time lock
+  return outboundEngine.translateSwap(
+    outboundAddress,
+    hash,
+    outboundAmount,
+    OUTBOUND_TIME_LOCK
+  )
+}
+
+/**
+ * Try to translate a swap cross-chain by retrieving the preimage from the
+ * downstream chain and returning it to the upstream chain, cancelling the
+ * upstream when an unrecoverable error is encountered.
+ *
+ * @private
+ * @param {string}  hash                  - Base64 string of the hash for the
+ *                                          swap
+ * @param {Payment} inboundPayment        - Expected inbound payment
+ * @param {Engine}  inboundPayment.engine - Engine of the expected inbound
+ *                                          payment
+ * @param {Payment} outboundPayment       - Outbound payment we will make to
+ *                                          retrieve the preimage.
+ * @returns {string}                        Base64 encoded preimage for the swap
+ */
+async function translateOnce (hash, inboundPayment, outboundPayment) {
+  const {
+    paymentPreimage,
+    permanentError
+  } = await getPreimage(hash, inboundPayment, outboundPayment)
+
+  if (permanentError) {
+    logger.error(permanentError)
+    logger.error('Downstream payment encountered an error, cancelling ' +
+      `upstream invoice for ${hash}`)
+    await inboundPayment.engine.cancelSwap(hash)
+
+    const err = new Error(permanentError)
```

Comment:
We have custom errors for other types and although there is no duplication with this specific error, would it be worth it to create `PermanentError` class?

Reply:
That could be a good idea.